core/src/main/java/org/apache/spark/api/shuffle/ShuffleWriteSupport.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.api.shuffle;
+
+import java.io.IOException;
+
+import org.apache.spark.annotation.Experimental;
+
+/**
+ * :: Experimental ::
+ * An interface for deploying a shuffle map output writer
+ *
+ * @since 3.0.0
+ */
+@Experimental
+public interface ShuffleWriteSupport {
```

Comment:
Since `ShuffleWrite(/Read)Support` only contains 1 function, why do we add this layer(Also, `xxSupport` is ambiguous)?
Could we make `ShuffleExecutorComponents.writes` return a `ShuffleMapOutputWriter` directly?

Reply:
Found another reason to remove this `Write/ReadSupport` layer, in `ReadSupport`, although not proposed yet : P, it straightway contains partition level functions: https://github.com/palantir/spark/blob/62c2664f1f298889357c6ebeb9b6f08962c94ceb/core/src/main/java/org/apache/spark/api/shuffle/ShuffleReadSupport.java#L31-L38. However for `WriteSupport` path this is what 1-level-lower layer would do(partition-level writer).

Reply:
It's important to keep interfaces minimal, and to keep each interface responsible for a single set of functionality. Since `ShuffleExecutorComponents` is eventually also going to support lifecycle operations - particularly `stopExecutor` - I'd like to keep that method separate from the methods that create the writers.

Reply:
I think there might be a world where we can coalesce `ShuffleWriteSupport` and `ShuffleReadSupport` into a more generic `ShuffleIO` - although we'd probably want to reconsider the naming of the interface at the root of the plugin tree, which is currently `ShuffleDataIO`.

Reply:
Do you mean that `ShuffleWriteSupport` can contain more semantics/fields than `ShuffleMapOutputWriter`? Can you name some of them? Maybe because the name `xxSupport` is ambiguous I think all in `xxSupport` can be put in `xxer`

Reply:
Sorry but could you elaborate on coalescing these 2 to 1? Because `InputStream` and `OutputStream` are completely different so I don't understand how to do that.

Reply:
I think I agree with @gczsjdy .  I don't see a good reason to separate ShuffleWriteSupport from ShuffleMapOutputWriter.  I think the extra separation just makes things more confusing.  (is there a time when you have one of these but not the other?  or 1 of (a) but N of (b)?  does it help with testing?  etc. etc.)  Unless you have a use case which takes advantage of the extra layers?

Reply:
I thought about this a bit more. One motivation to separation write APIs from read APIs is to pass in only the subsection of the plugin tree that is applicable in each case. So I only want to pass write-specific functionality to SortShuffleWriter, and only pass read-specific shuffle functionality to the reader side.

But I hold this conviction pretty loosely. We can change this - let me know.

Reply:
@mccheah Yeah seperating write from read is pretty natural. Here we think maybe remove the `WriteSupport` layer, only the `ShuffleMapOutputWriter` is enough, and on read side, the current `ReadSupport` can be replaced with `ShuffleBlockReader`

Reply:
That doesn't separate the concerns as I described. The only layer above `ShuffleWriteSupport` is `ShuffleExecutorComponents`. If we remove `ShuffleWriteSupport`, we make `ShuffleExecutorComponents` responsible for `createMapOutputWriter`. But presumably `ShuffleExecutorComponents` would also have `createPartitionReader`, meaning now `ShuffleExecutorComponents` passed to `BypassMergeSortShuffleWriter` will now have both read and write methods accessible in the writer code.

I would like `BypassMergeSortShuffleWriter` to only be able to call `createShuffleMapOutputWriter`, and nothing else.

Reply:
Can the `MapOutputWriter` be created before passing to `BypassMergeSortShuffleWriter`(`MapOutputWriter` as a construction parameter)?

Reply:
A comment left here @mccheah